### ISSUE TITLE
High memory allocation when using longer AckTimeout

### ISSUE THREAD

**Issue by adrianotr** [fsharplang-ru/pulsar-client-dotnet#288]
There seems to be a correlation between the ack timeout and memory allocation. The longer I set it, the more memory gets allocated.

Here you can see the allocations if the timeout is set to 100 days, but the problem is visible for smaller ack timeouts as well.

<img width="950" alt="Screenshot 2024-12-17 at 5 09 57 PM" src="https://github.com/user-attachments/assets/5e154cf7-46d4-4476-a877-4922c5920f54" />

You can use the program bellow to reproduce the issue:

```csharp
var client = await new PulsarClientBuilder()
    .ServiceUrl("pulsar://localhost:6650")
    .BuildAsync();

long memoryBefore = GC.GetTotalMemory(true);

await using var consumer = await client.NewConsumer()
    .Topic("test-topic")
    .SubscriptionName("test-subscription")
    .AckTimeout(TimeSpan.FromDays(100))
    .SubscribeAsync();

long memoryAfter = GC.GetTotalMemory(true);
Console.WriteLine($"Memory used: {(memoryAfter - memoryBefore) / 1_000_000} mb");
```

**Comment by adrianotr** [comment]
Btw, the reason why we discovered this issue is that if we have the consumer set as the default (Zero) it seems that the client library effectively sets the ack timeout to 30s (which is a discrepancy to the Java client that keeps it as infinite).

**Comment by Lanayx** [comment]
@adrianotr good notice! It indeed correlates well with [the code](https://github.com/fsprojects/pulsar-client-dotnet/blob/develop/src/Pulsar.Client/Internal/UnAckedMessageTracker.fs#L38-L41) which totally correlates [with Java code](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageTracker.java#L136-L139)

>  we discovered this issue

Not sure I understand what is the issue though

**Comment by adrianotr** [comment]
Hi, thanks for the quick response. Sorry for not being very clear; let me explain that better.

> Not sure I understand what is the issue though

It's a separate issue, but basically it's why we were using very long ack timeouts. If you don't set the ack timeout, then the lib sets the ack timeout to 30s instead of keeping it as zero (only when retries are enabled apparently). This is what I meant that is different from the Java client.


> It indeed correlates well with the code

So the problem there is not the correlation but the immense amount of memory used.
For example, this is the memory allocation I get by using different ack timeouts:

| AckTimeout  | Memory Allocation |
|-------------|-------------------|
| 30 seconds  | 3 mb              |
| 24 hours    | 15 mb             |
| 2 days      | 27 mb             |
| 10 days     | 132 mb            |
| 100 days    | 750 mb            |

**Comment by Lanayx** [comment]
> If you don't set the ack timeout, then the lib sets the ack timeout to 30s instead of keeping it as zero 

That's incorrect, Pulsar.Client follows Java and [sets timeout to zero](https://github.com/fsprojects/pulsar-client-dotnet/blob/develop/src/Pulsar.Client/Api/Configuration.fs#L95)

> So the problem there is not the correlation but the immense amount of memory used gets increased.

I've provided the code reference so you could understand _why_ the amount of memory used gets increased.

```fsharp
let stepsCount = Math.Ceiling(ackTimeout.TotalMilliseconds / tickDuration.TotalMilliseconds) |> int
for _ in [1..stepsCount] do
    timePartitions.Enqueue(RedeliverSet())
```
According to the Math, if you increase `ackTimeout`, the `timePartitions` queue will increase as well. You can reduce memory used by increasing `AckTimeoutTickTime` together with increasing `AckTimeout`

**Comment by Lanayx** [comment]
Having said that, I've just noticed, that you might see 30s because [of this configuration](https://github.com/fsprojects/pulsar-client-dotnet/blob/develop/src/Pulsar.Client/Api/ConsumerBuilder.fs#L16). And it [was removed in Java at some point](https://github.com/apache/pulsar/pull/19486), so you can create a PR with applying this improvement to Pulsar.Client as well.
